src/Fabulous/Memo.fs

```diff
@@ -0,0 +1,101 @@
+namespace Fabulous
+
+open Fabulous
+
+module Memo =
+
+    type internal MemoData =
+        {
+            /// Captures type of data that memoization depends on
+            KeyData: obj
+
+            // comparer that remembers KeyType internally
+            KeyComparer: obj -> obj -> bool
+
+            /// Lambda that users provide
+            CreateWidget: obj -> Widget
+
+            /// Captures type hash of data that memoization depends on
+            KeyTypeHash: int
+
+            /// Captures type hash of the marker memoized function produces
+            MarkerTypeHash: int
+        }
+
+    type Memoized<'t> = { phantom: 't }
+
+    let private MemoAttributeKey = AttributeDefinitionStore.getNextKey()
+    let internal MemoWidgetKey = WidgetDefinitionStore.getNextKey()
+
+    let inline private getMemoData (widget: Widget) : MemoData =
+        (widget.ScalarAttributes
+         |> Array.find(fun a -> a.Key = MemoAttributeKey))
```

Comment:
Good point, I'm not sure about it either. One thought I have is regarding `key` property in a collection. But, I guess, we can say that add it inside the `lazy` function.

Reply:
bummer, `lazy` is a keyword

Reply:
<img width="491" alt="image" src="https://user-images.githubusercontent.com/6633216/148635954-efc3103d-597e-4678-93a5-d887f19946c0.png">

Reply:
You're right. Depending on our implementation of key, this might change.
Right now, I would say we should keep `key` only applicable to `WidgetBuilder` and not to `Memoized`.
The memo node has only 1 single responsibility